src/porepy/models/solution_strategy.py

```diff
@@ -508,26 +516,78 @@ def check_convergence(
             # At least for the default direct solver, scipy.sparse.linalg.spsolve, no
             # error (but a warning) is raised for singular matrices, but a nan solution
             # is returned. We check for this.
-            diverged = bool(np.any(np.isnan(solution)))
+            diverged = bool(np.any(np.isnan(solution_increment)))
             converged: bool = not diverged
-            error: float = np.nan if diverged else 0.0
-            return error, converged, diverged
+            residual_error: float = np.nan if diverged else 0.0
+            increment_error: float = np.nan if diverged else 0.0
+            # TODO decide whether to omit logging of errors for linear problems
+            # return residual_error, increment_error, converged, diverged
         else:
             # First a simple check for nan values.
-            if np.any(np.isnan(solution)):
+            if np.any(np.isnan(solution_increment)):
                 # If the solution contains nan values, we have diverged.
-                return np.nan, False, True
-            # Simple but fairly robust convergence criterion. More advanced options are
-            # e.g. considering errors for each variable and/or each grid separately,
-            # possibly using _l2_norm_cell
-            #
-            # We normalize by the size of the solution vector.
-            # Enforce float to make mypy happy
-            error = float(np.linalg.norm(solution)) / np.sqrt(solution.size)
-            logger.info(f"Normalized residual norm: {error:.2e}")
-            converged = error < nl_params["nl_convergence_tol"]
+                return np.nan, np.nan, False, True
+
+            # Increment based error
+            increment_error = self.nonlinear_increment_error(solution_increment)
+            # Residual based error
+            residual_error = self.nonlinear_residual_error(residual, init_residual)
+            logger.info(f"Normalized increment error: {increment_error:.2e}")
+            logger.info(f"Normalized residual error: {residual_error:.2e}")
+            # Check convergence
+            converged_inc = increment_error < nl_params["nl_convergence_tol"]
+            # Allow for nan values in the residual error for effectively disabled
+            # convergence check.
+            converged_res = (
+                residual_error < nl_params["nl_convergence_tol_res"]
+                or nl_params["nl_convergence_tol_res"] is np.inf
+            )
```

Comment:
To explain: I managed to define characteristic functions which produce nan values (due to division by zero). Even when these in the end produce return 0 values, the residual still is nan. If this is confusing, I suggest removing these extra treatments from this PR. I just had them included as they have been part of active code development.